core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -237,601 +220,1206 @@ private SqlQueryScheduler(
-                .collect(toImmutableMap(PlanNode::getId, node -> getTableInfo(session, metadata, node)));
+        Optional<DistributedStagesScheduler> distributedStagesScheduler = createDistributedStagesScheduler(currentAttempt.get());
+        distributedStagesScheduler.ifPresent(scheduler -> distributedStagesSchedulingTask = executor.submit(scheduler::schedule, null));
     }
 
-    private static TableInfo getTableInfo(Session session, Metadata metadata, TableScanNode node)
+    private synchronized Optional<DistributedStagesScheduler> createDistributedStagesScheduler(int attempt)
     {
-        TableSchema tableSchema = metadata.getTableSchema(session, node.getTable());
-        TableProperties tableProperties = metadata.getTableProperties(session, node.getTable());
-        return new TableInfo(tableSchema.getQualifiedName(), tableProperties.getPredicate());
-    }
+        if (queryStateMachine.isDone()) {
+            return Optional.empty();
+        }
+        DistributedStagesScheduler distributedStagesScheduler = PipelinedDistributedStagesScheduler.create(
+                queryStateMachine,
+                schedulerStats,
+                nodeScheduler,
+                nodePartitioningManager,
+                stages,
+                coordinatorStagesScheduler,
+                executionPolicy,
+                failureDetector,
+                schedulerExecutor,
+                splitSourceFactory,
+                splitBatchSize,
+                dynamicFilterService,
+                tableExecuteContextManager,
+                retryPolicy,
+                attempt);
+        this.distributedStagesScheduler.set(distributedStagesScheduler);
+        distributedStagesScheduler.addStateChangeListener(state -> {
+            if (queryStateMachine.getQueryState() == QueryState.STARTING && state != DistributedStagesSchedulerState.PLANNED) {
+                queryStateMachine.transitionToRunning();
+            }
 
-    private static Map<StageId, Set<StageId>> getStageLineage(QueryId queryId, SubPlan planTree)
-    {
-        ImmutableMap.Builder<StageId, Set<StageId>> result = ImmutableMap.builder();
-        for (SubPlan planNode : Traverser.forTree(SubPlan::getChildren).breadthFirst(planTree)) {
-            result.put(
-                    getStageId(queryId, planNode.getFragment().getId()),
-                    planNode.getChildren().stream()
-                            .map(stage -> getStageId(queryId, stage.getFragment().getId()))
-                            .collect(toImmutableSet()));
-        }
-        return result.build();
+            if (state.isDone() && !state.isFailure()) {
+                stages.getDistributedStagesInTopologicalOrder().forEach(execution -> stages.get(execution.getStageId()).finish());
+            }
+
+            if (stages.getCoordinatorStagesInTopologicalOrder().isEmpty()) {
+                if (state == DistributedStagesSchedulerState.FINISHED) {
```

Comment:
Please add a comment why we need this. Specifically when it is possible for `stages.getCoordinatorStagesInTopologicalOrder().isEmpty` to be true.